gstd/src/lock/mutex.rs

```diff
@@ -226,22 +280,36 @@ impl<'a, T> Future for MutexLockFuture<'a, T> {
     // mutex can be taken, else it waits (goes into *waiting queue*).
     fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
         let current_msg_id = msg::id();
-        let lock = unsafe { &mut *self.mutex.locked.get() };
-        if lock.is_none() {
-            *lock = Some(current_msg_id);
-            Poll::Ready(MutexGuard {
-                mutex: self.mutex,
-                holder_msg_id: current_msg_id,
-            })
-        } else {
-            // If the message is already in the access queue, and we come here,
-            // it means the message has just been woken up from the waitlist.
-            // In that case we do not want to register yet another access attempt
-            // and just go back to the waitlist.
-            if !self.mutex.queue.contains(&current_msg_id) {
-                self.mutex.queue.enqueue(current_msg_id);
+        let current_block = exec::block_height();
+        let locked_by = unsafe { &mut *self.mutex.locked.get() };
+        if let Some((lock_owner_msg_id, deadline_block)) = *locked_by {
+            if current_block >= deadline_block {
+                if let Some(msg_future_task) = async_runtime::futures().get_mut(&lock_owner_msg_id)
+                {
+                    msg_future_task.set_lock_exceeded();
```

Comment:
what if I poll my own caught lock?

Reply:
Can you please provide more details? I do not follow your question

Reply:
let lock = Mutex::lock()
lock.poll(); lock.poll() (not with await, but manually by pinning future and providing context)

Reply:
Tricky question for me as I am too new to rust. What might be the point of doing polling manually instead of relying on the async runtime? Will all the other stuff (all the futures) work if one decides to go with manul polling?

Reply:
Others will work. Furthermore we have test that checks exactly these sync primitives in manual. That's pretty similar to consequent awaits

Reply:
anyway it could be checked other time. better to merge current pr already